File: epi_judge_java/epi/SumRootToLeaf.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class SumRootToLeaf {
    private static int totalSum = 0;

    @EpiTest(testDataFile = "sum_root_to_leaf.tsv")
    public static int sumRootToLeaf(BinaryTreeNode<Integer> tree) {
        totalSum = 0;
        if (tree == null) return totalSum;
        calc(tree, 0);
        return totalSum;
    }

    private static void calc(BinaryTreeNode<Integer> tree, int currSum) {
        if (tree.left == null && tree.right == null) {
            totalSum += currSum * 2 + tree.data;
            return;
        }

        if (tree.left != null) calc(tree.left, currSum * 2 + tree.data);
        if (tree.right != null) calc(tree.right, currSum * 2 + tree.data);
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "SumRootToLeaf.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/sum_root_to_leaf.cc
#include "binary_tree_node.h"
#include "test_framework/generic_test.h"

int calc(const unique_ptr<BinaryTreeNode<int>> &tree, int currSum);

int SumRootToLeaf(const unique_ptr<BinaryTreeNode<int>> &tree) {
    return calc(tree, 0);
}

int calc(const unique_ptr<BinaryTreeNode<int>> &tree, int currSum) {
    if (tree == nullptr) return 0;

    currSum = currSum * 2 + tree->data;

    if (tree->left == nullptr && tree->right == nullptr) return currSum;

    return calc(tree->left, currSum) + calc(tree->right, currSum);
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"tree"};
    return GenericTestMain(args, "sum_root_to_leaf.cc", "sum_root_to_leaf.tsv",
                           &SumRootToLeaf, DefaultComparator{}, param_names);
}
